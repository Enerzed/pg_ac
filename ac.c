/*
 * ac.c
 */


#include "ac.h"


/*
 * Init and fini
 */


/* Init automaton storage */ 
void _PG_init(void) 
{
    /* Init automaton storage is currently not called on extension load until _PG_fini actually works
     * ac_init();
     */
}


/* Finalize */ 
void _PG_fini(void)
{
    /* Fini automaton storage is currently not called on extension load until _PG_fini actually works
     * ac_init();
     */
}


/*
 * Memory management
 */


/* Init automaton storage */ 
static void _ac_init() 
{
    HASHCTL ctl;
    memset(&ctl, 0, sizeof(ctl));
    ctl.keysize = sizeof(int64);
    ctl.entrysize = sizeof(ac_automaton_entry);
    /* Use TopMemoryContext for global storage */
    ctl.hcxt = TopMemoryContext; 
    automaton_storage = hash_create("automaton storage", INITIAL_NELEM, &ctl, HASH_ELEM | HASH_CONTEXT);
}


/* Cleanup automatons */ 
static void _ac_fini() 
{
    HASH_SEQ_STATUS status;
    ac_automaton_entry *entry;

    if (automaton_storage == NULL)
        return;

    hash_seq_init(&status, automaton_storage);

    /*  Iterate through all entries and free associated automaton */
    while ((entry = (ac_automaton_entry *) hash_seq_search(&status)) != NULL)
    {
        ac_automaton *automaton = entry->automaton;
        if (automaton) 
        {
            /* Free the trie, TSVector, and automaton struct */
            ac_free_trie(automaton->root);
            MemoryContextDelete(entry->automaton->ctx);
            pfree(automaton);
        }
    }

    /* Destroy the hash table after all entries are processed */
    hash_destroy(automaton_storage);
    automaton_storage = NULL;
}


/* Free trie */
void ac_free_trie(ac_state* current)
{
	if (current == NULL)
	{
		return;
	}
	for (int i = 0; i < MAX_CHILDREN; i++)
	{
		if (current->children[i] != NULL)
		{
			ac_free_trie(current->children[i]);
		}
	}
	pfree(current);
}


/*
 * Aho Corasick functions
 */


/* Creates a new Aho Corasick state */
ac_state* ac_create_state()
{
    /* Allocate and initialize a new state */
	ac_state* state = (ac_state*)palloc0(sizeof(ac_state)); 
	state->is_final = false;
	state->index = -1;
    state->depth = 0;
	state->fail_link = NULL;
	return state;
}


/* Adds a keyword to the trie */ 
void ac_add_keyword(ac_state* root, const char* keyword, const int index)
{
	ac_state* current = root;
    for (int i = 0; keyword[i]!= '\0'; i++)
    {
        if (current->children[keyword[i]] == NULL)
        {
            current->children[keyword[i]] = ac_create_state();
        }
        current = current->children[keyword[i]];
        current->depth = i + 1;
    }
    current->is_final = true;
    current->index = index;
}


/* Builds failure and dictionary links */ 
void ac_build_links(ac_state* root)
{
	int queue_capacity= 16;
	int queue_size = 0;
	ac_state** queue = (ac_state**)palloc0(queue_capacity * sizeof(ac_state*));
	int front = 0, rear = 0;
	queue[rear++] = root;
	queue_size++;

	while(front < rear)
	{
		ac_state* current = queue[front++];
        for (int i = 0; i < MAX_CHILDREN; i++)
		{
			ac_state* child = current->children[i];
			if (child && current == root)
				child->fail_link = root;
			else if (child)
			{
				ac_state* fail = current->fail_link;
                while(fail && !fail->children[i])
					fail = fail->fail_link;
				if (fail)
				{
                    if (fail->is_final)
                        child->dictionary_link = fail->children[i];
					child->fail_link = fail->children[i];
				}
				else 
					child->fail_link = root;
			}
			if (child)
			{
				if (queue_size == queue_capacity)
				{
					queue_capacity *= 2;
					queue = (ac_state**)repalloc(queue, queue_capacity * sizeof(ac_state*));
				}
				queue[rear++] = child;
				queue_size++;
			}
		}
	}
	pfree(queue);
}


/* Matches a text against the trie and counts the number of matches */
ac_match_result ac_match(ac_state* root, char* text)
{
    ac_state* current = root;
    ac_match_result result = {0};
    int *matches = palloc(16 * sizeof(int));
    int *counts = palloc(16 * sizeof(int));
    int capacity = 16;
    
    for (int i = 0; i < text[i] != '\0'; i++) 
    {	
        unsigned char c = (unsigned char)text[i];
        while (current && !current->children[c])
        {
            current = current->fail_link;
        }
        current = current ? current->children[c] : root;

        for (ac_state *temp = current; temp; temp = temp->dictionary_link)
        {
            if (temp->is_final) 
            {
                if (result.num_matches == capacity) 
                {
                    capacity *= 2;
                    matches = repalloc(matches, capacity * sizeof(int));
                    counts = repalloc(counts, capacity * sizeof(int));
                }
                matches[result.num_matches] = temp->index;
                counts[result.num_matches] = 1;
                result.num_matches++;
            }
        }
    }

    result.matches = matches;
    result.counts = counts;
    return result;
}


/*
 * Searches for a given text in the trie 
 * (at least one match and not necessarily the same as in the text)
 * and immediately returns if found 
 */
bool ac_contains(ac_state *root, const char *text) 
{
    ac_state *current = root;
    for (int i = 0; text[i] != '\0'; i++)
    {
        unsigned char c = (unsigned char)text[i];
        while (current && !current->children[c])
        {
            current = current->fail_link;
        }
        current = current ? current->children[c] : root;
        
        for (ac_state *temp = current; temp; temp = temp->dictionary_link)
        {
            if (temp->is_final)
                return true;
        }
    }
    return false;
}


/* Helper function for ac_search (recursive) */
bool evaluate_query(QueryItem *item, TSQuery *tsq, ac_automaton *automaton) 
{
    /* If the item is a value */ 
    if (item->type == QI_VAL)            
    {
        char *lexeme;
        bool found;
        /* Get lexeme from TSQuery */
        lexeme = pnstrdup(GETOPERAND(tsq) + item->qoperand.distance, item->qoperand.length);  
        /* Look for that lexeme in the trie*/
        found = ac_contains(automaton->root, lexeme);                                          
        pfree(lexeme);
        return found;
    }
    /* If the item is an operator */
    else if (item->type == QI_OPR)   
    {
        switch (item->qoperator.oper) 
        {
            case OP_AND:
                return evaluate_query(item + item->qoperator.left, tsq, automaton) && evaluate_query(item + 1, tsq, automaton);
            case OP_OR:
                return evaluate_query(item + item->qoperator.left, tsq, automaton) || evaluate_query(item + 1, tsq, automaton);
            case OP_NOT:
                return !evaluate_query(item + item->qoperator.left, tsq, automaton);
            default:
                elog(ERROR, "unrecognized operator: %d", item->qoperator.oper);
        }
    }
    return false;
}


/*
 * PostgreSQL-specific functions
 */


/* Init Aho Corasick automaton storage */
Datum ac_init(PG_FUNCTION_ARGS)
{
    _ac_init();
    PG_RETURN_BOOL(true);
}


/* Fini Aho Corasick automaton storage */
Datum ac_fini(PG_FUNCTION_ARGS)
{
    _ac_fini();
    PG_RETURN_BOOL(true);
}


Datum ac_rank_simple(PG_FUNCTION_ARGS)
{
    int64 id;
    text *input;
    bool found;
    ac_automaton_entry *entry;
    char *text_str;
    ac_match_result result;
    float4 score;
    /* Get automaton id */ 
    id = PG_GETARG_INT64(0);
    /* Get text */
    input = PG_GETARG_TEXT_PP(1);
    /* Look for the automaton */ 
    entry = hash_search(automaton_storage, &id, HASH_FIND, &found);
    /* If not found, return false */ 
    if (!found)
        PG_RETURN_FLOAT4(0.0);
    /* Else, evaluate */ 
    text_str = text_to_cstring(input);
    result = ac_match(entry->automaton->root, text_str);
    score = (float)result.num_matches / entry->automaton->num_lexemes;
    /* Clean up */
    pfree(result.matches);
    pfree(result.counts);
    pfree(text_str);
    /* Return score */
    PG_RETURN_FLOAT4(score);
}


/* Build Aho Corasick automaton */
Datum ac_build(PG_FUNCTION_ARGS) 
{
    MemoryContext oldctx;;
    MemoryContext ctx;
    TSVector tsv;
    ac_automaton *automaton;
    WordEntry *entries;
    int64 id;
    ac_automaton_entry *entry;
    /**/
    ctx = AllocSetContextCreate(TopMemoryContext, "Automaton context", ALLOCSET_DEFAULT_SIZES);
    /* Set current memory context */ 
    oldctx = MemoryContextSwitchTo(ctx);
    /* Get TSVector */
    tsv = PG_GETARG_TSVECTOR_COPY(0);
    /* Create automaton */ 
    automaton = palloc0(sizeof(ac_automaton));
    automaton->num_lexemes = tsv->size;
    automaton->root = ac_create_state();
    automaton->ctx = ctx;
    /* Build trie */
    entries = ARRPTR(tsv);
    /* Get next automaton id*/
    id = next_automaton_id++;
    /* Add keywords to trie */
    for (int i = 0; i < tsv->size; i++) 
    {
        char *lexeme;
        lexeme = pnstrdup(STRPTR(tsv) + entries[i].pos, entries[i].len);
        ac_add_keyword(automaton->root, lexeme, i);
        pfree(lexeme);
    }
    /* Build links */
    ac_build_links(automaton->root);
    /* Store automaton */
    entry = hash_search(automaton_storage, &id, HASH_ENTER, NULL);
    entry->id = id;
    entry->automaton = automaton;
    /* Set old memory context and return */ 
    MemoryContextSwitchTo(oldctx);
    /* Return automaton id */
    PG_RETURN_INT64(id);
}


/* Destroy Aho Corasick automaton */
Datum ac_destroy(PG_FUNCTION_ARGS)
{
    int64 id;
    bool found;
    ac_automaton_entry *entry;
    /* Get automaton id*/
    id = PG_GETARG_INT64(0); 
    /* Look for the automaton*/
    entry = hash_search(automaton_storage, &id, HASH_FIND, &found);
    /* If not found, return false */
    if (!found)
        PG_RETURN_BOOL(false);
    /* Else, destroy */
    ac_free_trie(entry->automaton->root);
    MemoryContextDelete(entry->automaton->ctx);
    hash_search(automaton_storage, &id, HASH_REMOVE, NULL);
    /* Return true */
    PG_RETURN_BOOL(true);
}


/* Search in Aho Corasick automaton using TSQuery */
Datum ac_search_tsquery(PG_FUNCTION_ARGS) 
{
    int64 id;
    TSQuery tsq;
    bool found;
    ac_automaton_entry *entry;
    QueryItem *items;
    bool result = false;
    /* Get automaton id */
    id = PG_GETARG_INT64(0);
    /* Get TSQuery */
    tsq = PG_GETARG_TSQUERY(1);
    /* Look for the automaton */ 
    entry = hash_search(automaton_storage, &id, HASH_FIND, &found);
    /* If not found, return false */
    if (!found)
        PG_RETURN_BOOL(false);
    /* Else, evaluate */
    items = GETQUERY(tsq);
    result = evaluate_query(items, tsq, entry->automaton);
    /* Return result */
    PG_RETURN_BOOL(result);
}


/* Search in Aho Corasick automaton using text */
Datum ac_search_text(PG_FUNCTION_ARGS) 
{
    int64 id;
    text *input;
    bool found;
    ac_automaton_entry *entry;
    char *text_str;
    bool result = false;
    /* Get automaton id*/
    id = PG_GETARG_INT64(0);
    /* Get text */
    input = PG_GETARG_TEXT_PP(1);
    /* Look for the automaton */
    entry = hash_search(automaton_storage, &id, HASH_FIND, &found);
    /* If not found, return false */
    if (!found)
        PG_RETURN_BOOL(false);
    /* Else, evaluate */
    text_str = text_to_cstring(input);
    result = ac_contains(entry->automaton->root, text_str);
    pfree(text_str);
    /* Return result */
    PG_RETURN_BOOL(result);
}


/* TODO DONT USE */
Datum ac_match_text(PG_FUNCTION_ARGS) 
{
    int64 id = PG_GETARG_INT64(0);              // Get automaton id
    text *input = PG_GETARG_TEXT_PP(1);         // Get text
    char *text_str = text_to_cstring(input);
    ac_match_result result;

    // Look for the automaton
    bool found;
    ac_automaton_entry *entry = hash_search(automaton_storage, &id, HASH_FIND, &found);
    // If not found, return false
    if (!found)
        PG_RETURN_FLOAT4(0.0);
    // Else, evaluate
    result = ac_match(entry->automaton->root, text_str);

    pfree(result.matches);
    pfree(result.counts);
    pfree(text_str);
    
}