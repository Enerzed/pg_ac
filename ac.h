/*
 * ac.h
 */


#pragma once


#include "postgres.h"
#include "fmgr.h"

#include "utils/palloc.h"
#include "utils/memutils.h"
#include "utils/array.h"
#include "utils/lsyscache.h"
#include "utils/builtins.h"
#include "utils/hsearch.h"
#include "utils/jsonb.h"

#include "tsearch/ts_cache.h"
#include "tsearch/ts_locale.h"
#include "tsearch/ts_public.h"
#include "tsearch/ts_type.h"
#include "tsearch/ts_utils.h"

#include "access/tupdesc.h"
#include "access/htup_details.h" 

#include "catalog/pg_ts_dict.h"
#include "catalog/namespace.h"

#include "nodes/parsenodes.h"
#include "common/hashfn.h"
#include "libpq/pqformat.h"

#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>
#include <stdint.h>
#include <string.h>
#include <ctype.h>
#include <funcapi.h>


#define MAX_CHILDREN 256
#define MAX_LEXEME_SIZE 256
#define INITIAL_NELEM 256

PG_MODULE_MAGIC;


/* Aho Corasick trie node */
typedef struct
{
	/* Children of the node */
	struct ac_state* children[MAX_CHILDREN];
	/* Failure link traverses to the link that has the longest common prefix */
	struct ac_state* fail_link;
	/* Dictionary link traverses to the link that is also need to be considered */
	struct ac_state* dictionary_link;
	/* Index of the lexeme */
	int index;
	/* Depth of the node */
	int depth;
	/* Is the node contains a lexeme */
	bool is_final;
} ac_state;


/* Aho Corasick Automaton */
typedef struct 
{
	/* Trie */
	ac_state *root;
	/* Total number of lexemes */
    int num_lexemes;
} ac_automaton;


/* Aho Corasick automaton entry for storage */
typedef struct
{
	/* Automaton id */
	int32 id;
	/* Automaton */
	ac_automaton *automaton;
} ac_automaton_entry;


/* Useful for ranking only */
typedef struct 
{
	/* Match indices */
    int *matches;
	/* Number of matches for each match index, stored as 1 for each match */
    int *counts;
	/* Total number of matches */
    int num_matches;
} ac_match_result;


/* Global variables for automaton storage */
static HTAB *automaton_storage = NULL;
static int next_automaton_id = 1;


/* Init and fini */
void _PG_init(void);
void _PG_fini(void);


/*
 * Memory management 
 */
static void init_automaton_storage();
static void cleanup_automaton();
void ac_free_trie(ac_state* current);


/*
 * Aho Corasick functions 
 */

/* Create and initialize Aho Corasick state */
ac_state* ac_create_state();
/* Add keyword to the trie */
void ac_add_keyword(ac_state* root, const char* keyword, const int index);
/* Build failure and dictionary links */
void ac_build_links(ac_state* root);
/* Match indices */
ac_match_result ac_match(ac_state* root, char* text);
/* Look if the word is in the trie */
bool ac_contains(ac_state *root, const char *text);
/* Evaluate query (recursive)*/
bool evaluate_query(QueryItem *item, TSQuery *tsq, ac_automaton *automaton);


/* 
 * PostgreSQL-specific functions 
 */

/* Build Aho Corasick automaton */
Datum ac_build(PG_FUNCTION_ARGS);
/* Init Aho Corasick automaton storage */
Datum ac_init(PG_FUNCTION_ARGS);
/* Destroy all Aho Corasick automatons */
Datum ac_fini(PG_FUNCTION_ARGS);
/* Destroy Aho Corasick automaton */
Datum ac_destroy(PG_FUNCTION_ARGS);
/* Search in Aho Corasick automaton using TSQuery */
Datum ac_search_tsquery(PG_FUNCTION_ARGS);
/* Search in Aho Corasick automaton using text */
Datum ac_search_text(PG_FUNCTION_ARGS);
/* Match indices in Aho Corasick automaton using text */
Datum ac_match_text(PG_FUNCTION_ARGS);
/* Rank search result */
Datum ac_rank_simple(PG_FUNCTION_ARGS);

/*
 * PostgreSQL initialize functions 
*/
PG_FUNCTION_INFO_V1(ac_build);
PG_FUNCTION_INFO_V1(ac_init);
PG_FUNCTION_INFO_V1(ac_destroy);
PG_FUNCTION_INFO_V1(ac_fini);
PG_FUNCTION_INFO_V1(ac_search_tsquery);
PG_FUNCTION_INFO_V1(ac_search_text);
PG_FUNCTION_INFO_V1(ac_match_text);
PG_FUNCTION_INFO_V1(ac_rank_simple);