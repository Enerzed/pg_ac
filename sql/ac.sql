--TEST CASES--

/* Create extension */
CREATE EXTENSION pg_ac;
/* Init automaton storage */
SELECT ac_init();
/* Prepare test data */
SELECT ac_build(to_tsvector('english', 'Dogs and cats are the best pets'));
SELECT ac_build(to_tsvector('english', 'Dogs are the best pets'));
SELECT ac_build(to_tsvector('english', 'Cats are the best pets'));
SELECT ac_build(to_tsvector('english', 'Snakes are the best animals'));
SELECT ac_build(to_tsvector('english', 'Quick brown fox jumps over the lazy dog'));
SELECT ac_build(ARRAY['quick', 'brown', 'fox', 'jumps', 'over', 'the', 'lazy', 'dog']);

/* Test first automaton */
SELECT ac_search(1, to_tsquery('english', 'dog & cat'));                        -- Should be true
SELECT ac_search(1, to_tsquery('english', '!dog & !cat'));                      -- Should be false
SELECT ac_search(1, to_tsquery('english', 'dog | snake'));                      -- Should be true

/* Test second automaton */
SELECT ac_search(2, to_tsquery('english', 'pet & (cat | dog)'));                -- Should be true
SELECT ac_search(2, to_tsquery('english', 'pet & !(cat | dog)'));               -- Should be false
SELECT ac_search(2, to_tsquery('english', 'likely'));                           -- Should be false

/* Test third automaton */
SELECT ac_search(3, to_tsquery('english', 'pet & (dog | cat)'));                -- Should be true
SELECT ac_search(3, to_tsquery('english', 'pet & !(dog | cat)'));               -- Should be false
SELECT ac_search(3, to_tsquery('english', ''));                                 -- Should be false

/* Test fourth automaton */
SELECT ac_search(4, to_tsquery('english', 'are'));                              -- Should be false
SELECT ac_search(4, to_tsquery('english', 'pet & (dog | cat)'));                -- Should be false
SELECT ac_search(4, to_tsquery('english', 'the'));                              -- Should be false

/* Test fifth automaton for ranked search */
SELECT ac_rank_simple(5, 'jump dog fox');                                       -- Should be 0.5
SELECT ac_rank_simple(5,'fox');                                                 -- Should be 0.1(6)
SELECT ac_rank_simple(5,'pink horse');                                          -- Should be 0

/* Matches for fifth and sixth automaton, index structure is the same as corresponding TSVector indexes */
SELECT ac_match(5, 'quick quick jump dog fox');                            -- Should be 1,1,4,8,3
SELECT ac_match(6, 'quick quick jump dog fox');                            -- Should be 1,1,8,3
/* Clean up */
SELECT ac_fini();

DROP EXTENSION pg_ac;