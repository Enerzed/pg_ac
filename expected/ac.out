--TEST CASES--
/* Create extension */
CREATE EXTENSION pg_ac;
/* Init automaton storage */
SELECT ac_init();
 ac_init 
---------
 t
(1 row)

/* Prepare test data */
SELECT ac_build(to_tsvector('english', 'Dogs and cats are the best pets'));
 ac_build 
----------
        1
(1 row)

SELECT ac_build(to_tsvector('english', 'Dogs are the best pets'));
 ac_build 
----------
        2
(1 row)

SELECT ac_build(to_tsvector('english', 'Cats are the best pets'));
 ac_build 
----------
        3
(1 row)

SELECT ac_build(to_tsvector('english', 'Snakes are the best animals'));
 ac_build 
----------
        4
(1 row)

SELECT ac_build(to_tsvector('english', 'Quick brown fox jumps over the lazy dog'));
 ac_build 
----------
        5
(1 row)

/* Test first automaton */
SELECT ac_search(1, to_tsquery('english', 'dog & cat'));                        -- Should be true
 ac_search 
-----------
 t
(1 row)

SELECT ac_search(1, to_tsquery('english', '!dog & !cat'));                      -- Should be false
 ac_search 
-----------
 f
(1 row)

SELECT ac_search(1, to_tsquery('english', 'dog | snake'));                      -- Should be true
 ac_search 
-----------
 t
(1 row)

/* Test second automaton */
SELECT ac_search(2, to_tsquery('english', 'pet & (cat | dog)'));                -- Should be true
 ac_search 
-----------
 t
(1 row)

SELECT ac_search(2, to_tsquery('english', 'pet & !(cat | dog)'));               -- Should be false
 ac_search 
-----------
 f
(1 row)

SELECT ac_search(2, to_tsquery('english', 'likely'));                           -- Should be false
 ac_search 
-----------
 f
(1 row)

/* Test third automaton */
SELECT ac_search(3, to_tsquery('english', 'pet & (dog | cat)'));                -- Should be true
 ac_search 
-----------
 t
(1 row)

SELECT ac_search(3, to_tsquery('english', 'pet & !(dog | cat)'));               -- Should be false
 ac_search 
-----------
 f
(1 row)

SELECT ac_search(3, to_tsquery('english', ''));                                 -- Should be false
NOTICE:  text-search query doesn't contain lexemes: ""
 ac_search 
-----------
 f
(1 row)

/* Test fourth automaton */
SELECT ac_search(4, to_tsquery('english', 'are'));                              -- Should be false
NOTICE:  text-search query contains only stop words or doesn't contain lexemes, ignored
 ac_search 
-----------
 f
(1 row)

SELECT ac_search(4, to_tsquery('english', 'pet & (dog | cat)'));                -- Should be false
 ac_search 
-----------
 f
(1 row)

SELECT ac_search(4, to_tsquery('english', 'the'));                              -- Should be false
NOTICE:  text-search query contains only stop words or doesn't contain lexemes, ignored
 ac_search 
-----------
 f
(1 row)

/* Test fifth automaton for ranked search */
SELECT ac_rank_simple(5, 'jump dog fox');                                       -- Should be 0.5
 ac_rank_simple 
----------------
            0.5
(1 row)

SELECT ac_rank_simple(5,'fox');                                                 -- Should be 0.1(6)
 ac_rank_simple 
----------------
     0.16666667
(1 row)

SELECT ac_rank_simple(5,'pink horse');                                          -- Should be 0
 ac_rank_simple 
----------------
              0
(1 row)

/* Matches for fifth and sixth automaton, index structure is the same as corresponding TSVector indexes */
SELECT ac_match_text(5, 'quick quick jump dog fox');                            -- Should be 1,1,4,8,3
 ac_match_text 
---------------
 {1,1,4,8,3}
(1 row)

/* Clean up */
SELECT ac_fini();
 ac_fini 
---------
 t
(1 row)

DROP EXTENSION pg_ac;
